### PR TITLE
optimize aggregate status plugin for deployment

### PR DESCRIPTION
**Pull request by mrlihanbo**

### CODE REVIEW

File: pkg/controllers/statusaggregator/plugins/deployment.go [hunk 1]
```diff
@@ -45,16 +51,38 @@ func (receiver *DeploymentPlugin) AggregateStatues(
...
 	}
 
-	aggregatedStatus := &appsv1.DeploymentStatus{
-		ObservedGeneration: sourceDeployment.Generation,
+	aggregatedStatus := &appsv1.DeploymentStatus{}
+
+	resource := &fedtypesv1a1.GenericObjectWithStatus{}
+	err := util.UnstructuredToInterface(fedObject, resource)
+	if err != nil {
+		return nil, false, errors.Wrapf(err, "Failed to unmarshall to generic resource")
```
Comment (limhawjia) [inline pkg/controllers/statusaggregator/plugins/deployment.go:67]:
```suggestion
		return nil, false, fmt.Errorf("failed to unmarshall to generic resource: %w", err)
```

File: pkg/controllers/statusaggregator/plugins/deployment.go [hunk 2]
```diff
@@ -45,16 +51,38 @@ func (receiver *DeploymentPlugin) AggregateStatues(
...
+	}
+
+	templateSynced, err := checkResourceTemplateSyncStatus(sourceObject, fedObject, resource)
+	if err != nil {
+		logger.Error(err, fmt.Sprintf("Failed to check resource template %s/%s sync status",
```
Comment (limhawjia) [inline pkg/controllers/statusaggregator/plugins/deployment.go:72]:
```suggestion
		logger.Error(err, "Failed to check resource template sync status")
```

Comment (limhawjia) [reply pkg/controllers/statusaggregator/plugins/deployment.go:72]: The contextual logger will already have the object name

File: pkg/controllers/statusaggregator/plugins/deployment.go [hunk 3]
```diff
@@ -45,16 +51,38 @@ func (receiver *DeploymentPlugin) AggregateStatues(
...
+			sourceObject.GetNamespace(), sourceObject.GetName()))
+		return nil, false, err
+	}
+
+	if !templateSynced {
+		logger.V(3).Info(fmt.Sprintf("Resource template %s/%s is not synced",
```
Comment (limhawjia) [inline pkg/controllers/statusaggregator/plugins/deployment.go:78]:
```suggestion
		logger.V(3).Info("Resource template is not synced")
```

File: pkg/controllers/statusaggregator/plugins/deployment.go [hunk 4]
```diff
@@ -45,16 +51,38 @@ func (receiver *DeploymentPlugin) AggregateStatues(
...
+			sourceObject.GetNamespace(), sourceObject.GetName()))
+		needUpdateObservedGeneration = false
+	}
+
+	collectGenerations := make(map[string]int64, len(resource.Status.Clusters))
+	for _, cluster := range resource.Status.Clusters {
```
Comment (limhawjia) [inline pkg/controllers/statusaggregator/plugins/deployment.go:84]: `resource.Status` is a pointer field, we should add a nil check above

File: pkg/controllers/statusaggregator/plugins/deployment.go [hunk 5]
```diff
@@ -45,16 +50,38 @@ func (receiver *DeploymentPlugin) AggregateStatues(
...
+		return nil, false, err
+	}
+
+	if !templateSynced {
+		logger.V(3).Info("Resource template is not synced")
+		needUpdateObservedGeneration = false
+	}
+
+	collectGenerations := make(map[string]int64)
+	if resource.Status != nil {
```
Comment (limhawjia) [inline pkg/controllers/statusaggregator/plugins/deployment.go:81]: @gary-lgy If `resource.Status` is nil, does it imply that the federated object has not been synced once? If that is the case, can we check this above and terminate early?

File: pkg/controllers/statusaggregator/plugins/deployment.go [hunk 6]
```diff
@@ -103,6 +135,12 @@ func (receiver *DeploymentPlugin) AggregateStatues(
 		}
 	}
 
+	if needUpdateObservedGeneration {
```
Comment (limhawjia) [inline pkg/controllers/statusaggregator/plugins/deployment.go:138]: Let's add a comment to describe when `needUpdateObservedGeneration` will be true, and why we only update the observed generation when it is.

> We only update the source object's observed generation after it has been federated and synced, and we have aggregated the statuses of the latest cluster objects. It is only at this point where we can "successfully observe" the latest generation of the source object.

File: pkg/controllers/statusaggregator/plugins/deployment.go [hunk 7]
```diff
@@ -159,3 +197,152 @@ func (receiver *DeploymentPlugin) AggregateStatues(
 
 	return sourceObject, needUpdate, nil
 }
+
+func checkResourceTemplateSyncStatus(sourceObject, fedObject *unstructured.Unstructured,
```
Comment (gary-lgy) [inline pkg/controllers/statusaggregator/plugins/deployment.go:201]:
```suggestion
func checkResourceTemplateSyncStatus(
    sourceObject, fedObject *unstructured.Unstructured,
```

File: pkg/controllers/statusaggregator/plugins/deployment.go [hunk 8]
```diff
@@ -159,3 +197,152 @@ func (receiver *DeploymentPlugin) AggregateStatues(
...
+	prunedSourceJSON, err := jsonpatch.MergePatch(
+		sourceJSON,
+		mergePatch,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	var ret map[string]any
+	if err := utiljson.Unmarshal(prunedSourceJSON, &ret); err != nil {
```
Comment (gary-lgy) [inline pkg/controllers/statusaggregator/plugins/deployment.go:344]: Let's add a comment here documenting why `utiljson` is required and `json` doesn't work.

File: pkg/controllers/statusaggregator/plugins/deployment.go [hunk 9]
```diff
@@ -159,3 +197,152 @@ func (receiver *DeploymentPlugin) AggregateStatues(
...
+func checkResourceTemplateSyncStatus(sourceObject, fedObject *unstructured.Unstructured,
+	resource *fedtypesv1a1.GenericObjectWithStatus,
+) (bool, error) {
+	if sourceObject == nil {
+		return false, fmt.Errorf("source object can't be nil")
+	}
+
+	if fedObject == nil || resource == nil || resource.Status == nil {
+		return false, nil
+	}
```
Comment (gary-lgy) [inline pkg/controllers/statusaggregator/plugins/deployment.go:210]: These will never be nil, right? @limhawjia

File: pkg/controllers/statusaggregator/plugins/deployment.go [hunk 10]
```diff
@@ -45,16 +50,38 @@ func (receiver *DeploymentPlugin) AggregateStatues(
...
+
+	resource := &fedtypesv1a1.GenericObjectWithStatus{}
+	err := util.UnstructuredToInterface(fedObject, resource)
+	if err != nil {
+		return nil, false, fmt.Errorf("failed to unmarshall to generic resource: %w", err)
+	}
+
+	templateSynced, err := checkResourceTemplateSyncStatus(sourceObject, fedObject, resource)
+	if err != nil {
+		logger.Error(err, "Failed to check resource template sync status")
```
Comment (gary-lgy) [inline pkg/controllers/statusaggregator/plugins/deployment.go:71]: Errors returned from this method will be logged outside.

File: pkg/controllers/statusaggregator/plugins/deployment.go [hunk 11]
```diff
@@ -159,3 +197,152 @@ func (receiver *DeploymentPlugin) AggregateStatues(
 
 	return sourceObject, needUpdate, nil
 }
+
+func checkResourceTemplateSyncStatus(sourceObject, fedObject *unstructured.Unstructured,
+	resource *fedtypesv1a1.GenericObjectWithStatus,
```
Comment (gary-lgy) [inline pkg/controllers/statusaggregator/plugins/deployment.go:202]: `resource` seems unused.

File: pkg/controllers/statusaggregator/plugins/deployment.go [hunk 12]
```diff
@@ -45,16 +50,38 @@ func (receiver *DeploymentPlugin) AggregateStatues(
...
+		logger.Error(err, "Failed to check resource template sync status")
+		return nil, false, err
+	}
+
+	if !templateSynced {
+		logger.V(3).Info("Resource template is not synced")
+		needUpdateObservedGeneration = false
+	}
+
+	collectGenerations := make(map[string]int64)
```
Comment (gary-lgy) [inline pkg/controllers/statusaggregator/plugins/deployment.go:80]:
```suggestion
	clusterSyncedGenerations := make(map[string]int64)
```

File: pkg/controllers/statusaggregator/plugins/deployment.go [hunk 13]
```diff
@@ -79,6 +107,10 @@ func (receiver *DeploymentPlugin) AggregateStatues(
 			return nil, false, err
 		}
 
+		if collectGeneration, exist := collectGenerations[clusterName]; !exist || collectGeneration != deployStatus.ObservedGeneration {
```
Comment (gary-lgy) [inline pkg/controllers/statusaggregator/plugins/deployment.go:110]: Let's add a comment here.

File: pkg/controllers/util/propagationstatus/propagationstatus.go [hunk 14]
```diff
@@ -0,0 +1,168 @@
...
+		return synced, err
+	}
+
+	resource := &fedtypesv1a1.GenericObjectWithStatus{}
+	err = util.UnstructuredToInterface(fedObject, resource)
+	if err != nil {
+		return false, fmt.Errorf("failed to unmarshall to generic resource: %w", err)
+	}
+
+	templatePropagated := true
```
Comment (gary-lgy) [inline pkg/controllers/util/propagationstatus/propagationstatus.go:43]:
```suggestion
	syncAllOk := true
```

File: pkg/controllers/statusaggregator/plugins/deployment.go [hunk 15]
```diff
@@ -79,6 +98,11 @@ func (receiver *DeploymentPlugin) AggregateStatues(
 			return nil, false, err
 		}
 
+		// If the cluster's controller has not observed the latest synced generation, its status will be out-of-date.
+		if collectGeneration, exist := clusterSyncedGenerations[clusterName]; !exist || collectGeneration != deployStatus.ObservedGeneration {
```
Comment (gary-lgy) [inline pkg/controllers/statusaggregator/plugins/deployment.go:102]:
```suggestion
		if gen, exist := clusterSyncedGenerations[clusterName]; !exist || collectGeneration != deployStatus.ObservedGeneration {
```

**Pull request merged**